hlsl-headers/include/vk/khr/cooperative_matrix.impl

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "common.hlsli"
+
+namespace vk {
+namespace khr {
+namespace internal {
+
+// These enums are expected to match the values in spirv.h
+// TODO: See if there is a way to use a spirv header to get these values
+// instead of redefining them.
+enum CooperativeMatrixOperand {
+  NoneKHR = 0x0,
+  MatrixASignedComponentsKHR = 0x1,
+  MatrixBSignedComponentsKHR = 0x2,
+  MatrixCSignedComponentsKHR = 0x4,
+  MatrixResultSignedComponentsKHR = 0x8,
+  MatrixAllSignedComponents = 0xF,
+  SaturatingAccumulationKHR = 0x10,
+  SaturatingAllSignedComponenets = 0x1F
+};
+
+enum class CooperativeMatrixLayout {
+  RowMajorKHR = 0x0,
+  ColumnMajorKHR = 0x1,
+};
+
+enum class CooperativeMatrixUse {
+  MatrixAKHR = 0,
+  MatrixBKHR = 1,
+  MatrixAccumulatorKHR = 2,
+};
+
+// Type-Declaration Instructions
+
+// TODO: make sure scope, rows, and cols can be specialization constants
+template <typename ComponentType, uint scope, uint rows, uint columns, uint use>
+using CooperativeMatrixKHR = vk::SpirvOpaqueType<
+    /* OpTypeCooperativeMatrixKHR */ 4456, ComponentType,
+    vk::integral_constant<uint, scope>, vk::integral_constant<uint, rows>,
+    vk::integral_constant<uint, columns>, vk::integral_constant<uint, use> >;
+
+// Define the load and store instructions
+template <typename ResultType, typename PointerType>
+[[vk::ext_instruction(
+    /* OpCooperativeMatrixLoadKHR */
+    4457)]] [[SPV_KHR_CooperativeMatrix]] ResultType
```

Comment:
All instructions need `SPV_KHR_CooperativeMatrix`.

Reply:
This was wrong. The instruction do not need the capability and the extension. I've opted to have the capability on the member variable definition. This way, if you declare a variable of the class then the capability and extension will be added. None of the function are usable without adding an instance of the class. That should be enough. I'll need to modify the spec to allow the capability and extension on the member variable. It is already allowed on variables.